### DISCUSSION TITLE
Incompatibility between IPSEC policy DIR OUT and NAT ?

### DISCUSSION THREAD

**Discussion by carros06** [strongswan/strongswan#1022] [Q&A]
Hello
My system needs to support Device1 (Client) to Device2 (Serveur) communication through  IPSEC using NAT.
 IPSEC is managed by Strongswan  in MODULE1 and MODULE2  (see architecture.ppt attached file).
The MODULE1 IPSEC configuration is provided in module1ClientIPSEC.conf.
The NAT Forwarding rules on module1 (Client) is provided in module1Clientiptables.sh file
The IPSECstatus is provided in module1ClientIPSECstatusall.txt file
In order to prevent data flows that do not match the IPSEC rules to be sent in clear On Control Network (eg: data flow  Req to 172.169.12.150 from PC Windows) I put an IPSEC policy DIR OUT in the forwarding rule:
iptables -A FORWARD -i eth1 -o eth0 -p tcp --dport $rulePort -m policy --pol ipsec --dir out -j ACCEPT  
But this DIR OUT policy prevents the data flow using NAT to work correctly: Req to 192.169.1.80 is not sent on IPSEC by MODULE1
If I remove DIR OUT policy, the Data flow based on NAT is OK but Req to 172.169.12.150 from PC Windows is sent in clear by MODULE1.
What is the reason of this behavior ?
If No DIR OUT policy can be used, what can be done to avoid that Req to 172.169.12.150 from PC Windows is sent in clear by MODULE1 when the IPSEC rules are not matching ?

Many Thanks

[module1ClientIPSEC.txt](https://github.com/strongswan/strongswan/files/8583791/module1ClientIPSEC.txt)
[module1Clientiptables.txt](https://github.com/strongswan/strongswan/files/8583793/module1Clientiptables.txt)
[module1ClientIPSECstatusall.txt](https://github.com/strongswan/strongswan/files/8583797/module1ClientIPSECstatusall.txt)

[architecture.pptx](https://github.com/strongswan/strongswan/files/8583777/architecture.pptx)

**Comment by tobiasbrunner** [comment]
> iptables -A FORWARD -i eth1 -o eth0 -p tcp --dport $rulePort -m policy --pol ipsec --dir out -j ACCEPT

This is NOT an IPsec policy (those can be managed manually via `ip xfrm policy`). It's a firewall rule that uses the `policy` module to match packets against existing IPsec policies. In this case, the rule applies only to packets for which there is a matching outbound IPsec policy (and the other restrictions, like TCP and the port). And it has nothing really to do with NAT (rules for that are installed in `POSTROUTING` and the `nat` table).

You might want to read [this](https://docs.strongswan.org/docs/5.9/howtos/securityRecommendations.html#_tunnel_shunting).

**Comment by carros06** [comment]
Thanks for the information.
I tried to set up the following DROP policy in MODULE1 to avoid that Req from PC Windows (192.168.2.100) to DEVICE2 (172.169.12.150) is sent in clear by MODULE1 when the IPSEC rules are not matching
conn 192.169.1.80-tunnel-P2P-drop
    type=drop
    left=192.169.1.50
    leftsubnet=192.168.11.50/16
    right=192.169.1.80
    rightsubnet=0.0.0.0/0
    auto=route

Note: the righsubnet value is not known by MODULE1 configuration

It seems that this DROP policy disturbs the non IPSEC communications between DEVICE1: 192.168.10.1 and MODULE1: 192.168.11.50.

What could be done to avoid this disturbance ?

**Reply by tobiasbrunner** [reply]
With `rightsubnet=0.0.0.0/0` it will obviously block all traffic from `192.168.11.50/16` to any remote IP (except for IKE traffic, for which socket-specific bypass policies are installed). And as documented, you should not configure actual IP addresses for `left|right` in drop/bypass policies (set them to `127.0.0.1` instead). You need to add bypass policies for traffic you want to allow (or configure a more specific `rightsubnet` for the drop policy).

**Reply by carros06** [reply]
Could you have a look to my last post below ?
I think that I follow your recommendations and I still have a problem.

**Comment by carros06** [comment]
In order to solve the problem above, I add a passthrough policy to not disturb  the non IPSEC communications between DEVICE1: 192.168.10.1 and MODULE1: 192.168.11.50 and I change the DROP policy to the following:
(see attached updated file module1ClientIPSEC.txt)
[module1ClientIPSEC.txt](https://github.com/strongswan/strongswan/files/8601214/module1ClientIPSEC.txt)

conn 192.169.1.80-tunnel-P2P-drop
type=drop
left=127.0.0.1
leftsubnet=192.168.11.50/16
right=127.0.0.1
rightsubnet=0.0.0.0/0
auto=route

This DROP policy prevents the TUNNEL communications defined by the following rule to work:

conn 192.169.1.80-tunnel-P2P
    keyexchange=ikev2
    type=tunnel
    left=192.169.1.50
    leftsubnet=192.169.1.50/32
    right=192.169.1.80
    rightsubnet=192.169.1.80/32
    auto=start
    esp=null-sha256!

DROP Policy priority defined by 192.169.1.80-tunnel-P2P-drop is lower than TUNNEL policy defined by 192.169.1.80-tunnel-P2P.

So why DROP policy prevents TUNNEL policy to work ?

What can be done to have the TUNNEL policy working with the DROP policy ?

**Reply by tobiasbrunner** [reply]
> I add a passthrough policy to not disturb the non IPSEC communications between DEVICE1

What do you mean? Why install any drop policies then? Their general purpose is to drop non-IPsec traffic (to avoid plaintext leaks). You seem confused by these concepts. Maybe read [this](https://docs.strongswan.org/docs/5.9/howtos/introduction.html) and some background material on IPsec.

> This DROP policy prevents the TUNNEL communications defined by the following rule to work:

That's not very likely. However, your passthrough policy probably does because it has higher priority than regular IPsec policies, by default, and will exclude all traffic between `192.168.0.0/16` and `0.0.0.0/0` from IPsec processing.

**Reply by carros06** [reply]
Hello

Please have a look to architecture.ppt file to see the different needed data flows 
[architecture.pptx](https://github.com/strongswan/strongswan/files/8607837/architecture.pptx)


IPSEC configuration is provided in file:
[module1ClientIPSEC.txt](https://github.com/strongswan/strongswan/files/8607841/module1ClientIPSEC.txt)


Inside this file, you will find the following IPSEC rules:

192.169.1.80-passthroughbackplane:
This rule is to ensure that communication between DEVICE1 (@192.168.11.1) and MODULE1 (@192.168.11.50) is done outside IPSEC
(see data flow in green in file architecture.pptx)
conn 192.169.1.80-passthroughbackplane
    type=passthrough
    left=192.168.11.50
    right=192.168.11.1
    leftsubnet=192.168.11.50/16
    auto=route

192.169.1.80-tunnel-P2P:
P2P communications betwen DEVICE1 (@192.168.10.1)/MODULE1 (@192.169.1.50) and MODULE2 (@192.169.1.80)/DEVICE2 (@172.169.12.150)
 (see NAT data flow in blue in file architecture.pptx)
conn 192.169.1.80-tunnel-P2P
    keyexchange=ikev2
    type=tunnel
    left=192.169.1.50
    leftsubnet=192.169.1.50/32
    right=192.169.1.80
    rightsubnet=192.169.1.80/32
    auto=start
    esp=null-sha256!
This communication is using the following iptables rules in MODULE1:
[module1Clientiptables.txt](https://github.com/strongswan/strongswan/files/8608577/module1Clientiptables.txt)

192.169.1.80-tunnel-P2P-drop:
This DROP rule is to avoid PC (@192.168.2.100) or other device different from DEVICE1 on subnet 192.168.0.0 to send data in clear to any remote network through MODULE 1 (@192.168.11.50). 
(see data flow in red in file architecture.pptx)
conn 192.169.1.80-tunnel-P2P-drop
type=drop
left=127.0.0.1
leftsubnet=192.168.11.50/16
right=127.0.0.1
rightsubnet=0.0.0.0/0
auto=route

According to XFRM policy file below the priorities in MODULE1 are the following:
Data flow defined by 192.169.1.80-passthroughbackplane: priority 191807
Data flow defined by 192.169.1.80-tunnel-P2P: priority 367231
Data flow defined by 192.169.1.80-tunnel-P2P-drop: priority 591807
[module1xfrmPolicy.txt](https://github.com/strongswan/strongswan/files/8607919/module1xfrmPolicy.txt)


The problem is that the data flow defined by 192.169.1.80-tunnel-P2P is no more exchanged when the drop rule 192.169.1.80-tunnel-P2P-drop is added

I do not understand why because 192.169.1.80-tunnel-P2P-drop priority is lower than 192.169.1.80-tunnel-P2P.
How can I solve this problem ?

**Comment by carros06** [comment]
Hello,

Could you have a look to my last post ?
I think that I follow your recommendations and I still have a problem.

Many Thanks

A.Eyraud



Internal
De : Tobias Brunner ***@***.***>
Envoyé : lundi 2 mai 2022 11:47
À : strongswan/strongswan ***@***.***>
Cc : ANNICK EYRAUD ***@***.***>; Author ***@***.***>
Objet : Re: [strongswan/strongswan] Incompatibility between IPSEC policy DIR OUT and NAT ? (Discussion #1022)


[External email: Use caution with links and attachments]

________________________________



With rightsubnet=0.0.0.0/0 it will obviously block all traffic from 192.168.11.50/16 to any remote IP (except for IKE traffic, for which socket-specific bypass policies are installed). And as documented, you should not configure actual IP addresses for left|right in drop/bypass policies (set them to 127.0.0.1 instead). You need to add bypass policies for traffic you want to allow (or configure a more specific rightsubnet for the drop policy).

-

**Comment by carros06** [comment]
Hello,

Could you have a look to my last Post ?

Many Thanks

A.Eyraud

De : Tobias Brunner ***@***.***>
Envoyé : lundi 2 mai 2022 12:11
À : strongswan/strongswan ***@***.***>
Cc : ANNICK EYRAUD ***@***.***>; Author ***@***.***>
Objet : Re: [strongswan/strongswan] Incompatibility between IPSEC policy DIR OUT and NAT ? (Discussion #1022)


[External email: Use caution with links and attachments]

________________________________



I add a passthrough policy to not disturb the non IPSEC communications between DEVICE1

What do you mean? Why install any drop policies then? Their general purpose is to drop non-IPsec traffic (to avoid plaintext leaks). You seem confused by these concepts. Maybe read this<https://eur02.safelinks.protection.outlook.com/?url=https%3A%2F%2Fdocs.strongswan.org%2Fdocs%2F5.9%2Fhowtos%2Fintroduction.html&data=05%7C01%7Cannick.eyraud%40se.com%7C48cfcb5afca74bca93b308da2c2403cd%7C6e51e1adc54b4b39b5980ffe9ae68fef%7C0%7C0%7C637870830437676564%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000%7C%7C%7C&sdata=pnkrW1dQf%2Be345fk2bbwLg1Wv156eJsiPDYyLkG5RN8%3D&reserved=0> and some background material on IPsec.

This DROP policy prevents the TUNNEL communications defined by the following rule to work:

That's not very likely. However, your passthrough policy probably does because it has higher priority than regular IPsec policies, by default, and will exclude all traffic between 192.168.0.0/16 and 0.0.0.0/0 from IPsec processing.

-
